### PR TITLE
perf: use clones-with-immutable-args for QTokens, solmate instead of OZ for the base ERC20 and ERC1155 implementations, remove clutter as in duplicate code and checks

### PR DESCRIPTION
**Pull request by 0xca11**

### CODE REVIEW

File: spec/controller.spec [hunk 1]
```diff
@@ -210,6 +211,7 @@ invariant balanceVSsupply(uint collateralTokenId, address qToken, env e)
 */
 rule validQtoken(method f)  
 		filtered { f -> f.selector != certorafallback_0().selector &&
+						// f.selector != mintSpread(address,address,uint256).selector &&
```
[the remark this answers is not in the document]
Comment (0xca11) [reply spec/controller.spec]: Removed in a2c580a

File: spec/harness/CollateralTokenHarness.sol [hunk 2]
```diff
@@ -42,16 +42,16 @@ contract CollateralTokenHarness is CollateralToken {
...
-        public
-        view
-        returns (bool)
-    {
-        if (collateralTokenIds[i] == key) {
-            return true;
-        }
-        return false;
-    }
+    // function collateralTokenIdsContainsId(uint256 key, uint256 i)
```
[the remark this answers is not in the document]
Comment (0xca11) [reply spec/harness/CollateralTokenHarness.sol]: Removed in a2c580a

File: contracts/libraries/OptionsUtils.sol [hunk 3]
```diff
@@ -1,129 +1,121 @@
...
-                    _isCall
+    uint8 public constant STRIKE_PRICE_DECIMALS = 18;
+
+    /// @notice Splits a dinamically-sized byte array into an array of unsigned integers
+    /// in which each element represents a 32 byte chunk of the original data
+    /// @dev Uses the identity precompile to copy data from memory to memory
+    /// @param _data the original bytes to be converted to an array of uint256 values
+    /// @return result an array of uint256 values that represent 32 byte chunks of the original data,
+    /// in which the last byte represent the length of the original data
+    function bytesToUint256Array(bytes memory _data)
```
[the remark this answers is not in the document]
Comment (0xca11) [reply contracts/libraries/OptionsUtils.sol]: Simplified in 074f60c

File: contracts/options/CollateralToken.sol [hunk 4]
```diff
@@ -49,53 +59,50 @@ contract CollateralToken is ERC1155, ICollateralToken, EIP712, Ownable {
...
     function setOptionsFactory(address optionsFactory_) external onlyOwner {
         _optionsFactory = optionsFactory_;
     }
 
     /// @inheritdoc ICollateralToken
-    function createCollateralToken(
+    function createOptionCollateralToken(address _qTokenAddress)
+        external
+        override
+        onlyOwnerOrFactory
```
[the remark this answers is not in the document]
Comment (0xca11) [reply contracts/options/CollateralToken.sol]: Resolved in 380ad05

File: contracts/options/CollateralToken.sol [hunk 5]
```diff
@@ -49,53 +59,50 @@ contract CollateralToken is ERC1155, ICollateralToken, EIP712, Ownable {
...
+
+        emit CollateralTokenCreated(_qTokenAddress, address(0), id);
+    }
+
+    /// @inheritdoc ICollateralToken
+    function createSpreadCollateralToken(
         address _qTokenAddress,
         address _qTokenAsCollateral
-    ) external override returns (uint256 id) {
+    ) external override onlyOwnerOrFactory returns (uint256 id) {
```
[the remark this answers is not in the document]
Comment (0xca11) [reply contracts/options/CollateralToken.sol]: Resolved in 380ad05

File: test/Controller.test.ts [hunk 6]
```diff
@@ -1700,6 +1716,7 @@ describe("Controller", async () => {
       revertToSnapshot(snapshotId);
     });
 
+    // TODO: Rewrite this test using Forge
```
[the remark this answers is not in the document]
Comment (0xca11) [reply test/Controller.test.ts]: Added in 13316d8

File: test/QuantMath.test.ts [hunk 7]
```diff
@@ -145,62 +145,63 @@ describe("QuantMath lib", () => {
...
       assert.equal(
-        (await lib.testMul(a, a)).toString(),
+        (await lib.mulTest(a, a)).toString(),
         expectedResult.value.toString(),
         "multiplication result mismatch"
       );
     });
 
+    // TODO: rewrite this using Forge
     it("Should return overflow error when number is too big", async () => {
```
[the remark this answers is not in the document]
Comment (0xca11) [reply test/QuantMath.test.ts]: Added in e30eda1

File: test/CollateralToken.test.ts [hunk 8]
```diff
@@ -481,271 +470,14 @@ describe("CollateralToken", () => {
...
-
-  describe("getCollateralTokenInfo", () => {
-    it("Should revert when passing an invalid id", async () => {
-      await expect(
-        collateralToken.getCollateralTokenInfo(0)
-      ).to.be.revertedWith("CollateralToken: Invalid id");
-    });
-
-    it("Should return the correct info for non-spreads", async () => {
-      await createCollateralToken(
```
[the remark this answers is not in the document]
Comment (0xca11) [reply test/CollateralToken.test.ts:712]: Periphery contract with util functions added here: https://github.com/RollaProject/quant-periphery/commit/084bae2410ce4f7b0136f651b97bc5dedfdc5808

**Review by quantizations — APPROVED** [review]

**Pull request merged**